### PR TITLE
ERC-20 test

### PR DESCRIPTION
**Pull request by g-r-a-n-t**
### What was wrong?

We did not have a test validating the functionality of our ERC-20 contract.

### How was it fixed?

- Improved error formatting

`SemanticError` now has a function `with_context` that can be [used](https://github.com/g-r-a-n-t/fe/blob/e7b7af5e7833d87f2272b5ec2b9011448d5b96d8/semantics/src/traversal/expressions.rs#L40-L58) to attribute spans to errors so that we can print line numbers and surrounding code.

Example of an error encountered while attempting to compile the ERC-20 contract:

```
Error: TypeError on line 24
pub def __init__(_name: string100, _symbol: string100):
        self._name = _name
        self._symbol = _symbol
        self._decimals = 18
        self._mint(msg.sender, 2600)
```

*self._decimals = 18 is painted red on the terminal to indicate the precise location of the TypeError*

- implemented `pass` statements
- we are now adding user defined functions to the constructor code block so that we can `_mint` tokens during initialization 
- made some updates to the ERC-20 contract so that it would compile
- added contract tests for the ERC-20 contract

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] OPTIONAL: Update [Spec](https://github.com/ethereum/fe/blob/master/spec/index.md) if applicable
- [x] Clean up commit history

### CODE REVIEW

File: compiler/tests/compile_errors.rs [hunk 1]
```diff
@@ -53,14 +53,14 @@ use std::fs;
...
-        .expect("Unable to read fixture file");
-
-    match fe_compiler::evm::compile(&src, CompileStage::AllUpToBytecode) {
-        Err(CompileError { errors }) => assert_eq!(format!("{:?}", errors), error),
-        _ => panic!(
-            "Compiling succeeded when it was expected to fail with: {}",
-            error
-        ),
-    }
+    // let src = fs::read_to_string(format!("tests/fixtures/compile_errors/{}",
```
Comment (g-r-a-n-t) [inline compiler/tests/compile_errors.rs:56]: readd test

File: semantics/src/errors.rs [hunk 2]
```diff
@@ -1,16 +1,93 @@
...
+    }
+
+    pub fn not_callable() -> Self {
+        SemanticError {
+            kind: ErrorKind::NotCallable,
+            context: vec![],
+        }
+    }
+
+    pub fn with_context(mut self, span: Span) -> Self {
```
Comment (g-r-a-n-t) [inline semantics/src/errors.rs:89]: add doc

File: semantics/src/namespace/operations.rs [hunk 3]
```diff
@@ -82,7 +83,7 @@ mod tests {
         case(u256_bool_map(), U256_ARRAY_TYPE)
     )]
     fn type_error_index(value: Type, index: Type) {
-        let actual = operations::index(value, index).expect_err("didn't fail");
-        assert_eq!(actual, SemanticError::TypeError)
+        // let actual = operations::index(value, index).expect_err("didn't
```
Comment (g-r-a-n-t) [inline semantics/src/namespace/operations.rs:86]: readd test

File: semantics/src/traversal/assignments.rs [hunk 4]
```diff
@@ -117,10 +117,10 @@ mod tests {
 
     #[rstest(assignment, case("bar = 42"), case("self.foobar[42] = bar"))]
     fn type_error_assigns(assignment: &str) {
-        let result = analyze(scope(), assignment);
-        assert_eq!(
-            result.expect_err("semantic analysis did not return an error"),
-            SemanticError::TypeError
-        )
+        // let result = analyze(scope(), assignment);
```
Comment (g-r-a-n-t) [inline semantics/src/traversal/assignments.rs:120]: readd test

File: semantics/src/traversal/declarations.rs [hunk 5]
```diff
@@ -93,11 +93,11 @@ mod tests {
 
     #[test]
     fn type_error_decl() {
-        let statement = "foo: u256[100] = 26";
-        let result = analyze(scope(), statement);
-        assert_eq!(
-            result.expect_err("analysis didn't fail"),
-            SemanticError::TypeError
-        );
+        // let statement = "foo: u256[100] = 26";
```
Comment (g-r-a-n-t) [inline semantics/src/traversal/declarations.rs:96]: readd test

File: compiler/src/errors.rs [hunk 6]
```diff
@@ -41,6 +62,29 @@ impl CompileError {
             errors: vec![ErrorKind::Str(s)],
         }
     }
+
+    /// Creates a compile error from a semantic error and the source code.
+    pub fn from_semantic_error(error: SemanticError, src: &str) -> CompileError {
```
Comment (g-r-a-n-t) [inline compiler/src/errors.rs:67]: move this inside of SemanticError `format_with_src(&self, src) -> String`

File: compiler/src/yul/mappers/assignments.rs [hunk 7]
```diff
@@ -29,7 +29,8 @@ pub fn assign(
                 let target = expressions::expr(&context, first_target)?;
                 let value = expressions::expr(&context, value)?;
 
-                let typ: FixedSize = target_attributes.typ.to_owned().try_into()?;
+                let typ = FixedSize::try_from(target_attributes.typ.to_owned())
+                    .map_err(|_| CompileError::static_str("cant move"))?;
```
Comment (g-r-a-n-t) [inline compiler/src/yul/mappers/assignments.rs:33]: improve error

File: semantics/src/traversal/expressions.rs [hunk 8]
```diff
@@ -38,23 +38,24 @@ pub fn expr(
...
+        fe::Expr::CompOperation { .. } => expr_comp_operation(scope, Rc::clone(&context), exp),
+        fe::Expr::Call { .. } => expr_call(scope, Rc::clone(&context), exp),
         fe::Expr::List { .. } => unimplemented!(),
         fe::Expr::ListComp { .. } => unimplemented!(),
         fe::Expr::Tuple { .. } => unimplemented!(),
         fe::Expr::Str(_) => unimplemented!(),
         fe::Expr::Ellipsis => unimplemented!(),
-    };
+    }
+    .map_err(|err| err.with_context(exp.span))?;
```
Comment (g-r-a-n-t) [inline semantics/src/traversal/expressions.rs:58]: use `error`

File: src/main_full.rs [hunk 9]
```diff
@@ -93,7 +93,7 @@ fn compile(src_file: &str, output_dir: &str, targets: Vec<CompilationTarget>) ->
     let contents = fs::read_to_string(src_file)?;
 
     let output = fe_compiler::evm::compile(&contents, to_compile_stage(&targets))
-        .map_err(|e| Error::new(ErrorKind::Other, format!("{:?}", e)))?;
+        .map_err(|e| Error::new(ErrorKind::Other, format!("{}", e)))?;
```
Comment (g-r-a-n-t) [inline src/main_full.rs:96]: use `error`

File: src/main_full.rs [hunk 10]
```diff
@@ -69,7 +69,7 @@ pub fn main() {
 
     match compile(input_file, output_dir, targets) {
         Ok(_) => println!("Compiled {}. Outputs in {}", input_file, output_dir),
-        Err(err) => println!("Unable to compile {}. Error: {}", input_file, err),
+        Err(err) => println!("Unable to compile {}. \nError: {}", input_file, err),
```
Comment (g-r-a-n-t) [inline src/main_full.rs:72]: use `error`

File: semantics/src/errors.rs [hunk 11]
```diff
@@ -1,16 +1,122 @@
...
 }
+
+#[derive(Debug, PartialEq)]
+pub struct SemanticError {
+    pub kind: ErrorKind,
+    pub context: Vec<Span>,
+}
+
+impl SemanticError {
+    pub fn break_without_loop() -> Self {
```
Comment (g-r-a-n-t) [inline semantics/src/errors.rs:27]: add docs to these public functions

File: compiler/src/yul/mappers/contracts.rs [hunk 12]
```diff
@@ -39,28 +31,34 @@ pub fn contract_def(
...
             name: identifier! { Contract },
-            code: constructor::build(init),
+            code: constructor::build(init, runtime.clone()),
             objects: vec![yul::Object {
                 name: identifier! { runtime },
                 code: yul::Code {
-                    block: yul::Block { statements },
+                    block: yul::Block {
+                        statements: runtime,
+                    },
```
Comment (cburgdorf) [inline compiler/src/yul/mappers/contracts.rs:61]: I remember we talked before about this and you said that giving access to the user defined functions in the constructor would mean that we would have to duplicate each of these functions. Is this what we do now? If yes, then we should probably add a github issue to track a possible optimization (but maybe the YUL optimizer will optimize it away for us anyway...who knows)

File: compiler/src/yul/mod.rs [hunk 13]
```diff
@@ -18,7 +18,8 @@ pub struct CompilerOutput {
 pub fn compile(src: &str) -> Result<CompilerOutput, CompileError> {
     let tokens = fe_parser::get_parse_tokens(src)?;
     let fe_module = fe_parser::parsers::file_input(&tokens[..])?.1.node;
-    let context = fe_semantics::analysis(&fe_module)?;
+    let context = fe_semantics::analysis(&fe_module)
+        .map_err(|error| CompileError::str(error.format_with_src(src)))?;
```
Comment (cburgdorf) [inline compiler/src/yul/mod.rs:22]: Nice!

File: compiler/tests/evm_contracts.rs [hunk 14]
```diff
@@ -678,3 +701,106 @@ fn sized_vals_in_sto() {
...
+            vec![address_token(alice), address_token(james), uint_token(25)],
+            Some(bool_token(true)),
+        );
+        harness.test_function(
+            &mut executor,
+            "balanceOf",
+            vec![address_token(james)],
+            Some(uint_token(50)),
+        );
+    });
```
Comment (cburgdorf) [inline compiler/tests/evm_contracts.rs:805]: Man, it's so cool to see Fe being able to do all this! :tada:

File: semantics/src/errors.rs [hunk 15]
```diff
@@ -1,16 +1,131 @@
...
 }
+
+#[derive(Debug, PartialEq)]
+pub struct SemanticError {
+    pub kind: ErrorKind,
+    pub context: Vec<Span>,
+}
+
+impl SemanticError {
+    /// Create a new error with kind `BreakWithoutLooping`
```
Comment (cburgdorf) [inline semantics/src/errors.rs:27]: Nitpick `BreakWithoutLoop`~~`ing`~~

File: semantics/src/errors.rs [hunk 16]
```diff
@@ -1,16 +1,131 @@
...
+    pub fn break_without_loop() -> Self {
+        SemanticError {
+            kind: ErrorKind::BreakWithoutLoop,
+            context: vec![],
+        }
+    }
+
+    /// Create a new error with kind `ContinueWithoutLoop`
+    pub fn continue_without_loop() -> Self {
```
Comment (cburgdorf) [inline semantics/src/errors.rs:36]: You probably consider these two alternatives but still found the usage too clunky?

1. Write an `impl ErrorKind`

```
impl ErrorKind {

    pub fn to_semantic_error(self) -> SemanticError {
        SemanticError {
            kind: self,
            context: vec![],
        }
    }
}
```

Usage: `ErrorKind::ContinueWithoutLoop.to_semantic_error()`

2. 

```
impl SemanticError {

    pub fn from_error_kind(error_kind: ErrorKind) -> Self {
        SemanticError {
            kind: error_kind,
            context: vec![],
        }
    }
}

...
```

Usage: `SemanticError.from_error_kind(ErrorKind::ContinueWithoutLoop)`

**Review by cburgdorf — APPROVED** [review]
Really great to see this milestone achieved!

File: compiler/src/yul/mappers/contracts.rs (hunk 12 above)
Comment (g-r-a-n-t) [reply compiler/src/yul/mappers/contracts.rs:61]: > Is this what we do now?

yeah

> but maybe the YUL optimizer will optimize it away

Compiling the following without the optimizer

```javascript
object "Foo" {
  code {
    function bar() {
      mstore(0, 1)
    }

    mstore(2,3)
  }
}
```

results in

```
======= test.yul (EVM) =======

Pretty printed source:
object "Foo" {
    code {
        function bar()
        { mstore(0, 1) }
        mstore(2, 3)
    }
}


Binary representation:
600b565b60016000525b565b6003600252

Text representation:
    /* "test.yul":28:69   */
  jump(tag_1)
tag_2:
    /* "test.yul":61:62   */
  0x01
    /* "test.yul":58:59   */
  0x00
    /* "test.yul":51:63   */
  mstore
    /* "test.yul":43:69   */
tag_3:
  jump
tag_1:
    /* "test.yul":84:85   */
  0x03
    /* "test.yul":82:83   */
  0x02
    /* "test.yul":75:86   */
  mstore
```

whereas compiling with the optimizer results in

```
======= test.yul (EVM) =======

Pretty printed source:
object "Foo" {
    code { { mstore(2, 3) } }
}


Binary representation:
6003600252

Text representation:
    /* "test.yul":84:85   */
  0x03
    /* "test.yul":82:83   */
  0x02
    /* "test.yul":75:86   */
  mstore
```

so it looks like all of our unused functions will get removed.

File: compiler/src/yul/mappers/contracts.rs [hunk 17]
```diff
@@ -39,28 +31,34 @@ pub fn contract_def(
...
-        statements.push(runtime_abi::dispatcher(
-            attributes.public_functions.to_owned(),
-        )?);
+        let runtime = {
+            let mut runtime = runtime_functions::std();
+            runtime.append(&mut build_runtime_functions(
+                attributes.runtime_operations.to_owned(),
+            ));
+            runtime.append(&mut user_functions);
+            runtime.push(runtime_abi::dispatcher(
```
Comment (g-r-a-n-t) [inline compiler/src/yul/mappers/contracts.rs:47]: we shouldn't be adding the dispatcher to the constructor block

**Pull request merged**